sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -470,3 +397,451 @@ object JoinReorderDPFilters extends PredicateHelper {
+ */
+@DeveloperApi
+trait Crossover {
+
+  /**
+   * Generate a new [[Chromosome]] from the given parent [[Chromosome]]s,
+   * with this crossover algorithm.
+   */
+  def newChromo(father: Chromosome, mother: Chromosome) : Chromosome
+}
+
+/**
+ * This class implements the Genetic Edge Recombination algorithm. The algorithm generates
+ * a new traveling path by choosing the edges in certain order from the parents' edge table,
+ * where the edge table contains links of vertexes of the parents' traveling paths.
+ *
+ * Here's an example. Suppose we have two traveling paths on a graph,
+ *      I, [A B C D E F]
+ *     II, [B D C A E F]
+ *
+ * The algorithm works as follows,
+ *  1. find the links of each vertex, then we have an 'Edge Table'
+ *      A: B F C E
+ *      B: A C D F
+ *      C: B D A
+ *      D: C E B
+ *      E: D F A
+ *      F: A E B
+ *  2. from one vertex, say A, choose one of it's neighbours as it's new neighbour, say F, we have
+ *      {A F}
+ *  3. then choose one from F's neighbours(note that vertex that has been chosen should not
+ *     be chosen again),
+ *      {A F E}
+ *  4. go on,
+ *      {A F E D}
+ *  5. go on,
+ *      {A F E D B}
+ *  6. go on,
+ *      {A F E D B C}
+ *  Note if the procedure ends before all of vertexes been chosen, start from another vertex that
+ *  has not been chosen, and go on the procedure.
+ *
+ * For more information about the Genetic Edge Recombination,
+ * see "Scheduling Problems and Traveling Salesmen: The Genetic Edge
+ * Recombination Operator" by Darrell Whitley et al.
+ * https://dl.acm.org/citation.cfm?id=657238
+ */
+object EdgeRecombination extends Crossover {
+
+  def genEdgeTable(father: Chromosome, mother: Chromosome) : Map[JoinPlan, Seq[JoinPlan]] = {
+    val fatherTable = father.basicPlans.map(g => g -> findNeighbours(father.basicPlans, g)).toMap
+    val motherTable = mother.basicPlans.map(g => g -> findNeighbours(mother.basicPlans, g)).toMap
+
+    fatherTable.map(entry => entry._1 -> (entry._2 ++ motherTable(entry._1)))
+  }
+
+  def findNeighbours(genes: Seq[JoinPlan], g: JoinPlan) : Seq[JoinPlan] = {
+    val genesIndexed = genes.toIndexedSeq
+    val index = genesIndexed.indexOf(g)
+    val length = genes.size
+    if (index > 0 && index < length - 1) {
+      Seq(genesIndexed(index - 1), genesIndexed(index + 1))
+    } else if (index == 0) {
+      Seq(genesIndexed(1), genesIndexed(length - 1))
+    } else if (index == length - 1) {
+      Seq(genesIndexed(0), genesIndexed(length - 2))
+    } else {
+      Seq()
+    }
+  }
+
+  override def newChromo(father: Chromosome, mother: Chromosome): Chromosome = {
+    var newGenes: Seq[JoinPlan] = Seq()
+    // 1. Generate the edge table.
+    var table = genEdgeTable(father, mother)
+    // 2. Choose a start vertex randomly from the heads of father/mother.
+    var current =
+      if (util.Random.nextInt(2) == 0) father.basicPlans.head else mother.basicPlans.head
+    newGenes :+= current
```

Comment:
don't do this... This will create new `Seq`, which is consistently O(N) complexity.

Reply:
Fixed. Has switched to `mutable.Buffer` to finish the task.